### PR TITLE
Enforce Groupby.__iter__ deprecation and miscellaneous pytest fixes

### PR DESCRIPTION
**Pull request by galipremsagar**
## Description

This PR:

- [x] Enforces deprecation in `GroupBy.__iter__` 
- [x] Fixes miscellaneous pytest failures due to already existing differences in cudf vs pandas & nuly introduced `inferred_type` in Index.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/dataframe.py [hunk 1]
```diff
@@ -5860,9 +5861,11 @@ def _reduce(
             )
             source = self._get_columns_by_label(numeric_cols)
             if source.empty:
-                return Series(index=self.index)
-
-        axis = source._get_axis_from_axis_arg(axis)
+                return Series(
+                    index=cudf.Index([], dtype="str")
```
Comment (mroeschke) [inline python/cudf/cudf/core/dataframe.py:5865]: Curious why it should be `dtype="str"` and not `dtype=self.index.dtype`?

Comment (galipremsagar) [reply python/cudf/cudf/core/dataframe.py]: Yes, it should be `self.columns.dtype`. I fixed it now 👍

File: python/cudf/cudf/core/dataframe.py [hunk 2]
```diff
@@ -5860,9 +5861,11 @@ def _reduce(
             )
             source = self._get_columns_by_label(numeric_cols)
             if source.empty:
-                return Series(index=self.index)
-
-        axis = source._get_axis_from_axis_arg(axis)
+                return Series(
+                    index=self._data.to_pandas_index()[:0]
```
Comment (mroeschke) [inline python/cudf/cudf/core/dataframe.py:5865]: The `[:0]` is to get the empty `Index` with the correct type?

Comment (galipremsagar) [reply python/cudf/cudf/core/dataframe.py:5865]: Yup

**Review by mroeschke — APPROVED** [review]

**Pull request merged**